### PR TITLE
[FIXED JENKINS-36123] Add an extension point that allows plugins to veto polling

### PR DESCRIPTION
**Pull request by stephenc**
[JENKINS-36123](https://issues.jenkins-ci.org/browse/JENKINS-36123)

@jenkinsci/code-reviewers /cc @reviewbybees

### CODE REVIEW

File: core/src/main/java/jenkins/scm/SCMPollingDecisionHandler.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    public abstract boolean shouldPoll(Item item);
+
+    /**
+     * All registered {@link SCMPollingDecisionHandler}s
+     */
+    public static ExtensionList<SCMPollingDecisionHandler> all() {
+        return ExtensionList.lookup(SCMPollingDecisionHandler.class);
+    }
+
+}
```
Comment (andresrc) [inline core/src/main/java/jenkins/scm/SCMPollingDecisionHandler.java]: The loop logic is repeated in almost all use cases, maybe a helper method `public static boolean shouldPollItem(Item item)`  would be useful.

**Comment by oleg-nenashev** [comment]
Seems it has some overlap with [SCMCheckoutStrategy](http://javadoc.jenkins-ci.org/?jenkins/scm/SCMCheckoutStrategy.html).

I would rather vote for having a single DecisionHandler extension point for Checkout and Poll operations. So my proposal would be to generalize the new extension point a bit

File: test/src/test/java/hudson/triggers/SCMTriggerTest.java [hunk 2]
```diff
@@ -81,6 +87,29 @@ public void simultaneousPollAndBuild() throws Exception {
         build.get();  // let mock build finish
     }
 
+    /**
+     * Make sure that SCMTrigger doesn't trigger another build when a build has just started,
```
Comment (jtnord) [inline test/src/test/java/hudson/triggers/SCMTriggerTest.java]: This test does not seem to me to be actually checking what is described here.

File: core/src/main/java/hudson/triggers/SCMTrigger.java [hunk 3]
```diff
@@ -549,6 +553,16 @@ public void run() {
             if (job == null) {
                 return;
             }
+            // we can pre-emtively check the SCMPollingDecisionHandler instances here
+            // note that job().poll(listener) should also check this
+            for (SCMPollingDecisionHandler handler : SCMPollingDecisionHandler.all()) {
+                if (!handler.shouldPoll(job)) {
+                    LOGGER.log(Level.FINE, "Skipping polling for {0} due to veto from {1}",
```
Comment (jtnord) [inline core/src/main/java/hudson/triggers/SCMTrigger.java]: should this not go in the listener, so that it is available in the SCM polling log?

**Comment by stephenc** [comment]
@oleg-nenashev the problem is `SCMCheckoutStrategy` is exclusive to `AbstractProject` wherease pipelines are not in that hierarchy.

**Comment by jtnord** [comment]
some inline comments/questions are holing me back from a thumbs up - but it's not a thumbs down as it looks mostly good to me...

**Comment by jtnord** [comment]
:+1: / :bee:

**Comment by andresrc** [comment]
:bee:

**Comment by stephenc** [comment]
https://issues.jenkins-ci.org/browse/JENKINS-36147 is a related follow-up enhancement

**Comment by oleg-nenashev** [comment]
> @oleg-nenashev the problem is SCMCheckoutStrategy is exclusive to AbstractProject wherease pipelines are not in that hierarchy. 

My proposal would be to just rename <code>SCMPollingDecisionHandler</code> to something more generic like <code>SCMDecisionHandler</code>. Other code LGTM, I definitely do not propose to hack the Strategy extension point (At least because it behaves like a Job configuration component).

**Comment by andresrc** [comment]
:bee:

**Comment by oleg-nenashev** [comment]
🐝

File: core/src/main/java/jenkins/triggers/SCMTriggerItem.java [hunk 4]
```diff
@@ -116,6 +120,11 @@
                 return delegate.asProject().scheduleBuild2(quietPeriod, null, actions);
             }
             @Override public PollingResult poll(TaskListener listener) {
+                SCMDecisionHandler veto = SCMDecisionHandler.firstShouldPollVeto(asItem());
+                if (!veto.shouldPoll(asItem())) {
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/triggers/SCMTriggerItem.java]: @stephenc This is missing a `null` check, resulting in [JENKINS-36232](https://issues.jenkins-ci.org/browse/JENKINS-36232). Could you PTAL?

Comment (jtnord) [inline core/src/main/java/jenkins/triggers/SCMTriggerItem.java]: @daniel-beck already fixed? #2425

Comment (daniel-beck) [inline core/src/main/java/jenkins/triggers/SCMTriggerItem.java]: @jtnord Thanks. It seems the commit comment bot got broken by the JIRA update, so there are no closings nor commit references…

**Pull request merged**